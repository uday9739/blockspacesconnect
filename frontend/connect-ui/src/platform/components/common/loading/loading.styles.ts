import styled, { Keyframes, keyframes } from "styled-components";

/** Shimmer is an animation, moves the background with a gradient from left to right over a long distance in a short period of time. */
const shimmer: Keyframes = keyframes`
      0% {
        background-position: -1000px 0; width: 80%;
      }
      50%{
        width: 78%;
      }
      100% {
        background-position: 1000px 0; width: 80%
      }
`;

/**
 * #### SC is an animated styled component for the Loading component.
 *
 * @param $style
 * - An optional css style object.
 * - It is auto generated when any supported style attributes are passed to the Loading component.
 * - It has a '$' prefix, which makes it a transient attribute and it will not show up on the DOM.
 * - Its values will overwrite the default values.
 *
 * @remarks has two animations:
 * - Expands from 0 to 100% width.
 * - Takes 100% of the available space by default.
 * - Second animation is in the ::after pseudo element.
 * - Is a thin horizontal panel, height: 10px, width: 80% of the main div.
 * - Has a background with gradient thats moving side to side to create an effect of a shimmer.
 *
 * @example
 * ```js
 *  import {SC} from './';
 *  const Loading = ({ when, children, ...style }:LoadingProps) => {
 *    return when ? <SC.Loading role="presentation" $style={style}/> : children || null;
 *  };
 * ```
 * @see maxWidth prop passed to Loading component:
 * - Don't set 'width' directly, it will break the expand effect of Loading, set the width through maxWidth.
 * - maxWidth controls how wide it can stretch, the default is 100%, which means it will take 100% of the width the parent allows.
 * - All use cases are working with it now. if there is any exotic configuration we can set the dimensions and the position of the loader explicitly.
 * - Height, top, left, right, maxWidth, maxHeight, minWidth and maxHight can be set explicitly.
 * - The props can be set as numbers, then they default to pixels.
 * - If you want to use any other valid CSS unit, set them as strings.
 */
// SC is an animated styled component for the Loading component.
export const SC = {
  Loading: styled.div<{ $style?}>`
    display: flex;
    position: relative;
    justify-content: center;
    align-items: center;
    width: 100%;
    height: 100%;
    min-height: 1rem;
    margin: 0 auto;
    margin-bottom: 1.5rem;
    margin-top: 1.5rem;
    user-select: none;
    pointer-events: none;
    animation: expand 0.5s forwards cubic-bezier(0.25, 0.46, 0.45, 0.94);
    @keyframes expand {
      0% {
        width: 0;
      }
      100% {
        width: 100%;
      }
    }
    &::after {
      opacity: 0.8;
      position: absolute;
      width: 80%;
      height: 0.625rem;
      margin: 0 auto;
      border-radius: 0.525rem;
      content: "";
      background: linear-gradient(to right, #eff1f3 4%, #e2e2e2 25%, #eff1f3 36%);
      background-size: 62.5rem 100%;
      animation: ${shimmer} 2s infinite linear;
    }
    ${(props) => props.$style}
  `
};
