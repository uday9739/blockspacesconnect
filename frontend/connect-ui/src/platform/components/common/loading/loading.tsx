import React from "react";
import { SC } from './loading.styles';

/**
 * #### Props for the Loading component
 *
 * @remarks The css style properties defined in this type are available to be set.
 * - Additional css style properties for the Loading component can be added to this type.
 * - All style properties will automatically propagate to the styled component in a css style object.
 *
 * @example 1 (number will be converted to pixels)  or '1px' or '1rem' or '1em'.
 *
 * @see {@link SC} for implementation.
 */
export type LoadingProps = {
  /** - A prop that takes a function returning a boolean.
   *  - It will show and hide the wrapped component based on the return value of that function. */
  when: boolean;
  /** - Sets the height of the loading container
   *  - When you like for it to take up more vertical space */
  height?: string | number;
  /** - Sets the margins around the loading container.
   *  - When you like for it to take up more space on all sides */
  margin?: string | number;
  /** - Sets the position of the top edge of the loading container.
   *  - When the standard position (0) is not working in a use case.  */
  top?: string | number;
  /** - Sets the horizontal position of the loading container.
   *  - When the standard position (0) is not working in a use case.  */
  left?: string | number;
  /** - Sets the horizontal position of the right edge loading container.
   *  - When the standard position is not working in a use case.  */
  right?: string | number;
  /** - Sets the vertical position of the loading container.
   *  - When the standard position (0) is not working in a use case.  */
  bottom?: string | number;
  /** - Sets the maximum width of the loading container.
   *  - Setting this is a way to control the width.
   *  - Don't set width directly, it will break the opening animation.  */
  maxWidth?: string | number;
  /** - Sets the maximum height of the loading container */
  maxHeight?: string | number;
  /** - Sets the minimum width of the loading container */
  minWidth?: string | number;
  /** - Sets the minimum height of the loading container */
  minHeight?: string | number;
  /** - This is the wrapped component, the children nodes.
   *  - Will be set automatically */
  children?: any;
};


/**
 * #### Loading is a component that can be wrapped around any other component and show or hide it based on the boolean value passed through the 'when' prop.
 * - If 'when' prop is \{true\} render a loading animation.
 * - Else render the wrapped component.
 *
 * @param when - A boolean value that determines whether the loading component should hide the wrapped component or not.
 * @param maxWidth - optional
 * - Sets the final width of the loading animation after expansion.
 * - Overrides the standard 100%.
 * - Type is number or string.
 * - Number will convert to px.
 * - Use string to set valid css units -\> 1rem, 1px, 1em, 1% etc.
 * - Don't set 'width' directly, it will break the expanding animation of Loading, set the width through maxWidth.
 *
 * @remarks A loading component that can be used to display a loading indicator while data is being fetched from the server.
 *
 * @example Integration of the loading component into a layout.
 * ```js
 * import {Loading} from 'components/shared/loading'
 * import {Button} from 'components/shared/button'
 * <Loading when={true}><Button label="hidden by loading"/></Loading>
 * <Loading when={false}><Button label="revealed by loading"/></Loading>
 * ```
 * @see {@link LoadingProps} for other currently supported props.
 */
export const Loading = ({ when, children, ...style }: LoadingProps) => {
  return when ? <SC.Loading role="presentation" $style={style}/> : (children || <></>);
};