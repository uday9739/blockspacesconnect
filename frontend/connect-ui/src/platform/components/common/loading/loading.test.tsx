import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Loading } from './loading';
import "jest-styled-components";
import renderer from "react-test-renderer";

describe("testing loading components hide and reveal feature", () => {
  test("it renders a correct snapshot", async () => {
    const tree = renderer.create(<Loading when={true} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("has rendered the correct component", () => {
    render(
      <div role="main">
        <Loading when={true} top="25px"></Loading>
      </div>
    );
    expect(screen.getByRole("presentation")).toHaveStyleRule("width: 15px");
  });
  test("should render <Loading /> when loading is set to true", async () => {
    const { container, getByTestId } = render(
      <div role="main">
        <Loading when={true} top="25px"></Loading>
      </div>
    );
    expect(screen.getByRole("presentation")).toBeInTheDocument();
  });

  test("should not render <Loading /> when loading is set to false, without a child; the function should return null", () => {
    const { container } = render(
      <div role="main">
        <Loading when={false} top="25px"></Loading>
      </div>
    );

    expect(screen.queryByRole("presentation")).toBeNull();
  });

  test("<Loading />  should be rendered when the boolean is set to true and should then remove itself from the DOM when the boolean changes to false, and the inner text does the opposite", async () => {
    const { rerender } = render(
      <div role="main">
        <Loading when={true} top="25px">
          hello
        </Loading>
      </div>
    );
    expect(screen.getByRole("presentation")).toBeInTheDocument();
    expect(screen.queryByText("hello")).toBeNull();
    // re-render the same component with loading set to false
    rerender(
      <div role="main">
        <Loading when={false} top="25px">
          hello
        </Loading>
      </div>
    );
    expect(screen.getByRole("main")).toBeInTheDocument();
    expect(screen.queryByRole("presentation")).toBeNull();
    expect(screen.getByText("hello")).toBeInTheDocument();
  });

  test("<Loading />  should hide the wrapped component, then un-hide it, then hide it again, then un-hide it again", async () => {
    const { rerender } = render(
      <div role="main">
        <Loading when={true}>
          <button title="here is the button" />
        </Loading>
      </div>
    );
    expect(screen.getByRole("presentation")).toBeInTheDocument();
    expect(screen.queryByRole("button")).toBeNull();
    // re-render the same component with loading set to false
    rerender(
      <div role="main">
        <Loading when={false}>
          <button title="here is the button" />
        </Loading>
      </div>
    );
    expect(screen.queryByRole("presentation")).toBeNull();
    expect(screen.getByRole("button")).toBeInTheDocument();
    // re-render the same component with loading set to true
    rerender(
      <div role="main">
        <Loading when={true}>
          <button title="here is the button" />
        </Loading>
      </div>
    );
    expect(screen.getByRole("presentation")).toBeInTheDocument();
    expect(screen.queryByRole("button")).toBeNull();
    // re-render the same component with loading set to false
    rerender(
      <div role="main">
        <Loading when={false}>
          <button title="here is the button" />
        </Loading>
      </div>
    );
    expect(screen.queryByRole("presentation")).toBeNull();
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  test("<Loading />  should not let its wrapped component child be rendered when the boolean is set to true", async () => {
    const { container } = render(
      <div role="main">
        <Loading when={true} top="25px">
          <div data-testid="loading-child">Loading child</div>
        </Loading>
      </div>
    );
    expect(screen.queryByTestId("loading-child")).toBeNull();
  });
});

describe("testing loading components positioning, styles and dimensions", () => {
  test("the styles of the unmodified component should be correct", async () => {
    render(
      <div role="main" style={{ position: "absolute", width: 1000, height: 700 }}>
        <Loading when={true}>
          <button title="here is the button" />
        </Loading>
      </div>
    );
    const instanceLoading = screen.queryByRole("presentation");

    expect(instanceLoading).toHaveStyle("width: 100%");
    expect(instanceLoading).toHaveStyle("height: 100%");
    expect(instanceLoading).toHaveStyle("left: ");
    expect(instanceLoading).toHaveStyle("max-width: ");
    expect(instanceLoading).toHaveStyleRule("display: flex");
    expect(instanceLoading).toHaveStyleRule("height", "100%");
    expect(instanceLoading).toHaveStyleRule("align-items", "center");
    expect(instanceLoading).toHaveStyleRule("justify-content", "center");
    expect(instanceLoading).toHaveStyleRule("pointer-events", "none");
    expect(instanceLoading).toHaveStyleRule("user-select", "none");

    const rect = instanceLoading?.getBoundingClientRect();
    expect(rect?.left).toBe(0);
    expect(rect?.height).toBe(0);
    expect(rect?.width).toBe(0);
    expect(rect?.top).toBe(0);

    expect(instanceLoading).toHaveStyleRule("width", "80%", {
      modifier: "&::after"
    });
    expect(instanceLoading).toHaveStyleRule("background-size", "62.5rem 100%", {
      modifier: "&::after"
    });
  });

  test("should apply maxWidth correctly when set", async () => {
    render(
      <div role="main" style={{ position: "absolute", width: 1000, height: 700 }}>
        <Loading when={true} maxWidth="300">
          <button title="here is the button" />
        </Loading>
      </div>
    );

    const instanceLoading = screen.queryByRole("presentation");

    expect(instanceLoading).toHaveStyle("max-width: 300");
    expect(instanceLoading).toHaveStyleRule("max-width", "300");
  });

  test("should apply height and left correctly when set", async () => {
    render(
      <div role="main" style={{ position: "absolute", width: 1000, height: 700 }}>
        <Loading when={true} height="300" left="3rem">
          <button title="here is the button" />
        </Loading>
      </div>
    );

    const instanceLoading = screen.queryByRole("presentation");

    expect(instanceLoading).toHaveStyle("height: 300");
    expect(instanceLoading).toHaveStyle("left: 3rem");

    expect(instanceLoading).toHaveStyleRule("height", "300");
    expect(instanceLoading).toHaveStyleRule("left", "3rem");

    const rect = instanceLoading?.getBoundingClientRect();
    expect(rect?.left).toBe(0);
    expect(rect?.height).toBe(0);
    expect(rect?.x).toBe(0); // not sure why its not computing height here
    expect(rect?.width).toBe(0);
  });

  test("should accept provided position", async () => {
    const { rerender, queryByRole } = render(
      <div role="main" style={{ position: "absolute", width: 1000, height: 700 }}>
        <Loading when={true} left="100px" top="200px">
          <button title="here is the button" />
        </Loading>
      </div>
    );

    const instanceLoading = screen.queryByRole("presentation");

    expect(instanceLoading).toHaveStyle("left: 100px");
    expect(instanceLoading).toHaveStyle("top: 200px");
    expect(instanceLoading).toHaveStyleRule("left", "100px");
    expect(instanceLoading).toHaveStyleRule("top", "200px");
  });
});
