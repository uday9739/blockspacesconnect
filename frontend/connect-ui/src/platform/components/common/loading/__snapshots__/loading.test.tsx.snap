// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing loading components hide and reveal feature it renders a correct snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  height: 100%;
  min-height: 1rem;
  margin: 0 auto;
  margin-bottom: 1.5rem;
  margin-top: 1.5rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  -webkit-animation: expand 0.5s forwards cubic-bezier(0.25,0.46,0.45,0.94);
  animation: expand 0.5s forwards cubic-bezier(0.25,0.46,0.45,0.94);
}

.c0::after {
  opacity: 0.8;
  position: absolute;
  width: 80%;
  height: 0.625rem;
  margin: 0 auto;
  border-radius: 0.525rem;
  content: "";
  background: linear-gradient(to right,#eff1f3 4%,#e2e2e2 25%,#eff1f3 36%);
  background-size: 62.5rem 100%;
  -webkit-animation: ZzRnL 2s infinite linear;
  animation: ZzRnL 2s infinite linear;
}

<div
  className="c0"
  role="presentation"
/>
`;
